Hasura DDN Data Connector in TypeScript Course (#1019)
hasura/learn-graphql: tutorials/backend/hasura-v3-ts-connector/tutorial-site/content/aggregates.md

            const aggregate = request.query.aggregates[aggregateName];
            switch(aggregate.type) {
                case 'star_count':
                  
                case 'column_count':
                  
                case 'single_column':
            }
        }
    }
    
}
```

For now, we'll handle the first two cases here, and save the last for when we talk about custom aggregates.

In the first case, we want to generate a target list element which uses the `COUNT` SQL aggregate function.

```typescript
// ...
case 'star_count':
    target_list.push(`COUNT(1) AS ${aggregateName}`);
    break;
// ...
```

In the second case, we'll also use the `COUNT` function, but this time, we're counting non-null values in a single column:

```typescript
// ...
case 'column_count':
    target_list.push(`COUNT(${aggregate.column}) AS ${aggregateName}`);
    break;
// ...
```

We also need to interpret the `distinct` property of the aggregate object, and insert the `DISTINCT` keyword if needed:

```typescript
// ...
case 'column_count':
    target_list.push(`COUNT(${aggregate.distinct ? 'DISTINCT ' : ''}${aggregate.column}) AS ${aggregateName}`);
    break;
// ...
```

> Now let's update our generated SQL to use the generated target list:
> 
> ```typescript
> // ...
> const sql = `SELECT ${target_list.join(", ")} FROM (
>                (
>                  SELECT * FROM ${request.collection} ${where_clause} ${order_by_clause} ${limit_clause} ${offset_clause}
>                )`;
> // ...
> ```

Reviewer: At this point, we don't have a `sql` variable within `fetch_aggregates()`, so a user (read: me) might modify their `sql` variables that's already present and be confused.

```suggestion
We'll create a new generated SQL function within `fetch_aggregates()` to use the generated target list:

```typescript
// ...
const sql = `SELECT ${target_list.join(", ")} FROM (
               (
                 SELECT * FROM ${request.collection} ${where_clause} ${order_by_clause} ${limit_clause} ${offset_clause}
               )`;
// ...
```
```

Author: Done

                    throw new NotSupported("custom aggregates not yet supported");
            }
        }
    }

    const parameters: any[] = [];

    const limit_clause = request.query.limit == null ? "" : `LIMIT ${request.query.limit}`;
    
    const offset_clause = request.query.offset == null ? "" : `OFFSET ${request.query.offset}`;

    const where_clause = request.query.where == null ? "" : `WHERE ${visit_expression(parameters, request.query.where)}`;

    const order_by_clause = request.query.order_by == null ? "" : `ORDER BY ${visit_order_by_elements(request.query.order_by.elements)}`;

    const sql = `SELECT ${target_list.join(", ")} FROM (
                    SELECT * FROM ${request.collection} ${where_clause} ${order_by_clause} ${limit_clause} ${offset_clause}
                )`;

    console.log(JSON.stringify({ sql, parameters }, null, 2));

    const result = state.db.get(sql, ...parameters);

    if (result === undefined) {
        throw new InternalServerError("Unable to fetch aggregates");
    }

    return result;
}
```

That's it, so let's test our connector one more time, and hopefully see some passing tests this time. And let's 

> check that we're generating the right SQL.

Reviewer: I didn't mention this on the `predicates` page, but let's be explicit and tell them they need to kill and recompile their connector before running the tests.
